scripts/kb.py

```diff
@@ -0,0 +1,633 @@
+            .fetchone()["count(*)"]
+        )
+
+        # Build Annoy index in batches.
+        for row_id in tqdm.tqdm(
+            # We select by ROWID, which starts at 1.
+            range(1, row_count + 1, batch_size),
+            desc="Inferring entity embeddings",
+            position=0,
+        ):
+            ids = tuple(
+                (row["id"], row["ROWID"])
+                for row in self._db_conn.cursor()
+                .execute(
+                    f"""
+                        SELECT
+                            id,
+                            ROWID
+                        FROM
+                            entities
+                        WHERE
+                            ROWID BETWEEN {row_id} AND {row_id + batch_size - 1}
+                        ORDER BY
+                            ROWID
+                        """
+                )
+                .fetchall()
+            )
+            qids = tuple(_id[0] for _id in ids)
+            entities = load_entities(
+                language=self._language, qids=qids, db_conn=self._db_conn
+            )
+
+            # Assemble descriptions to be embedded.
+            ent_descs = [
+                " ".join({entities[qid].name, *entities[qid].aliases})
+                + " "
+                + (
+                    entities[qid].description
+                    if entities[qid].description
+                    else (
+                        entities[qid].article_text[:500]
+                        if entities[qid].article_text
+                        else ""
+                    )
+                )
+                for qid in qids
+            ]
+
+            for row_id_offset, qid, desc_vector in zip(
+                range(len(qids)),
+                qids,
+                [
+                    ent_desc_doc.vector
+                    for ent_desc_doc in nlp.pipe(texts=ent_descs, n_process=n_jobs)
+                ],
+            ):
+                self._annoy.add_item(
+                    # Annoy expects index to start with 0, so we index each entities vector by its entities.ROWID value
+                    # in the database shifted by -1.
+                    row_id + row_id_offset - 1,
+                    desc_vector
+                    if isinstance(desc_vector, numpy.ndarray)
+                    else desc_vector.get(),
+                )
+
+        logger.info("Building ANN index.")
+        self._annoy.build(n_trees=self._n_trees, n_jobs=n_jobs)
+
+    def get_candidates_all(
+        self, mentions: Iterator[Iterable[Span]]
+    ) -> Iterator[Iterable[Iterable[Candidate]]]:
+        """
+        Retrieve candidate entities for specified mentions per document. If no candidate is found for a given mention,
+        an empty list is returned.
+        mentions (Iterator[Iterable[Span]]): Mentions per documents for which to get candidates.
+        YIELDS (Iterator[Iterable[Iterable[Candidate]]]): Identified candidates per document.
```

Comment:
This type hurts my brain a little

Reply:
I get that :smile: I can decompose this into types if it's easier to read. The inner-most `Iterable[Candidate]` represents all entity candidates for a given mention. `Iterable[Iterable[Candidate]]` then represents all entity candidates for all mentions in a given document. The wrapping `Iterator` then indicates that you can iterate over the entity candidate sets document by document instead of having to everything in memory at once. Do you think it'd help to extend the docstring?

Reply:
Yea maybe some sort of type alias is useful here?

Reply:
Let me know if https://github.com/explosion/wikid/pull/4/commits/c62731f1910e95abb1426fa076a56b267d5b6c37 helps.